refactor(parser): replace peek in `parse_rest_binding` with checkpoint

camchenry opened the pull request:
- towards https://github.com/oxc-project/oxc/issues/11194

This removes lookahead usage in favor of checkpoints. I've also aligned the diagnostic for having a trailing comma after a rest element with the TypeScript error code and message.

camchenry commented:
* **#11225** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11225?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11225?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=11225).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented on crates/oxc_parser/src/diagnostics.rs:
```diff
@@ -217,8 +217,9 @@ pub fn spread_last_element(span: Span) -> OxcDiagnostic {
-pub fn binding_rest_element_trailing_comma(span: Span) -> OxcDiagnostic {
-    OxcDiagnostic::error("Unexpected trailing comma after rest element").with_label(span)
+pub fn rest_parameter_or_binding_pattern_may_not_have_trailing_comma(span: Span) -> OxcDiagnostic {
+    ts_error("1013", "A rest parameter or binding pattern may not have a trailing comma.")
```
This is also a normal js error, so we shouldn't display this as ts specific (ts error code).

Boshen commented:
### Merge activity

* **May 22, 4:44 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

camchenry committed:
refactor(parser): replace peek in `parse_rest_binding` with checkpoint (#11225)

- towards https://github.com/oxc-project/oxc/issues/11194

This removes lookahead usage in favor of checkpoints. I've also aligned the diagnostic for having a trailing comma after a rest element with the TypeScript error code and message.
